tracer/src/Datadog.Trace/Ci/TraceProcessors/OriginTagTraceProcessor.cs

```diff
@@ -0,0 +1,72 @@
+// <copyright file="OriginTagTraceProcessor.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using Datadog.Trace.Ci.Tags;
+using Datadog.Trace.Logging;
+using Datadog.Trace.TraceProcessors;
+
+namespace Datadog.Trace.Ci.TraceProcessors
+{
+    internal class OriginTagTraceProcessor : ITraceProcessor
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<TruncatorTraceProcessor>();
+
+        private readonly bool _isPartialFlushEnabled = false;
+        private readonly bool _isAgentlessEnabled = false;
+
+        public OriginTagTraceProcessor(bool isPartialFlushEnabled, bool isAgentlessEnabled)
+        {
+            _isPartialFlushEnabled = isPartialFlushEnabled;
+            _isAgentlessEnabled = isAgentlessEnabled;
+
+            Log.Information("OriginTraceProcessor initialized.");
+        }
+
+        public ArraySegment<Span> Process(ArraySegment<Span> trace)
+        {
+            // We ensure there's no trace (local root span) without a test tag.
+            // And ensure all other spans have the origin tag.
+
+            // Check if the trace has any span
+            if (trace.Count == 0)
+            {
+                // No trace to write
+                return trace;
+            }
+
+            if (!_isPartialFlushEnabled)
+            {
+                // Check if the last span (the root) is a test, bechmark or build span
+                Span lastSpan = trace.Array[trace.Offset + trace.Count - 1];
```

Comment:
I think this is an assumption we've made elsewhere too, but I'm not sure if it's always correct. For example, I'm seeing this behaviour with the Grpc integration, but due to the way threading works there, sometimes a child span is closed _after_ the root is. I haven't checked the implications, but maybe worth looking into?

Reply:
You're right, I've changed the algorithm to search for the root span in the trace in [1190e65](https://github.com/DataDog/dd-trace-dotnet/pull/2402/commits/1190e65c9e322803ea238766b66836134271f48a)